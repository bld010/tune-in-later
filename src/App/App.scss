.App {
  & header {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    padding: 20px;
  }

  & h1 {
    font-size: 50px
  }

  & nav {
    display: flex;
    flex-direction: row;
    justify-content: flex-end;

    & .nav {
      margin: 10px;
    }
  }
}



