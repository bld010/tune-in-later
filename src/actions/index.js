export const setCurrentUser = user => ({
  type: 'SET_CURRENT_USER',
  user
});

export const setCurrentUserFavorites = favorites => ({
  type: 'SET_CURRENT_USER_FAVORITES',
  favorites
});

export const displaySearchResults = searchResults => ({
  type: 'DISPLAY_SEARCH_RESULTS',
  searchResults

})

export const addUserFavorite = favorite => ({
  type: 'ADD_USER_FAVORITE',
  favorite
})