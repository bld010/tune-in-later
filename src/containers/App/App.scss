.App {
  // background: linear-gradient(0.25turn, #F29A93, #F7E1DA, #F29A93);
  // background: linear-gradient(0.25turn, #F7E1Da, #FFF, #F7E1DA);

  & header {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    padding: 20px;
  }

  & .main-display {
    min-height: 40vh;
    justify-content: center;
    align-items: center;
  }

  & h1 {
    font-size: 50px;
    margin: 0px auto;
    color: #666888;

    & span {
      font-weight: bold;
    }

  }

  & nav {
    display: flex;
    flex-direction: row;
    justify-content: center;
    width: 245px;
    margin: 10px 0px 0px 0px;

    & .nav {
      margin: 10px;
    }
  }
}



