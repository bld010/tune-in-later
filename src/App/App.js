import React, { Component } from 'react';
import BooksDisplay from '../containers/BooksDisplay/BooksDisplay';
import LoginForm from '../containers/LoginForm/LoginForm';
import { landingFetch, authorFetch } from '../util/apiCalls';
import './App.scss';
import NewUserForm from '../containers/NewUserForm/NewUserForm';
import { Route, NavLink, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { setCurrentUser } from '../actions';
import SearchForm from '../containers/SearchForm/SearchForm';
import BookDetails from '../containers/BookDetails/BookDetails'

class App extends Component{
  constructor() {
    super();
    this.state = {
      allBooks : [],
      error: '',
    }
  }
  
  componentDidMount() {
    landingFetch()
      .then(data => this.setState({allBooks: this.handleInitialData(data)}, () => {console.log(data)}))
      .catch(err => this.setState({error : 'Sorry, there was a problem loading our suggested audiobooks. Please enter a search term to see specific results!'}, () => {console.error('error in landing fetch', err)}))
  }

  handleInitialData = (data) => {
    const allBooks = data.map(datum => datum).flat();
    const formattedBooks = allBooks.map(book => ({
      artist: book.artistName,
      image: book.artworkUrl100,
      price: book.collectionPrice,
      title: book.collectionName,
      genre: book.primaryGenreName,
      description: book.description,
      date: book.releaseDate,
      filterType: book.filterType,
      id: book.collectionId
    }));
    return formattedBooks
  }

  filterAllBooks = (filterType) => {
    const filteredBooks = this.state.allBooks.filter(book => book.filterType === filterType);
    return filteredBooks;
  }

  render() { 
    return (
    <div className="App">
      <header>
        <SearchForm />
        <h1>ListenLater</h1>
        <nav>
          <NavLink to='/' className='nav'>Home</NavLink>
          {!this.props.currentUser && <NavLink to='/login' className='nav'>Sign In</NavLink> }
          {this.props.currentUser && <NavLink to='/' className='nav' onClick={() => this.props.setCurrentUser(null)}>Sign Out</NavLink>}
          {this.props.currentUser ? <h2 className='welcome-message'>Welcome {this.props.currentUser.name}</h2> : <h2>Welcome, please sign in!</h2>}
          {/* {this.props.currentUser && !this.props.favorites.length && <h2>You haven't favorited any books yet!</h2>} */}
        </nav>
      </header>

        <Route exact path='/login' render={() => {
          return(
            <>
              <LoginForm />
              <NewUserForm />
            </>)
          }} 
          />
            
        

      <Route 
        path = '/'
        render={() => {
          

            
          
          return(
            <>

          <section className="main-display">
            
            {this.props.selectedBook && <BookDetails book={this.props.selectedBook}/>}
          </section>


          {this.props.searchResults.length && <BooksDisplay 
            books={this.props.searchResults}
            sectionGenre='Search Results' />}
          {this.state.allBooks.length && <BooksDisplay 
            books={this.filterAllBooks('romance')} 
            sectionGenre='Romances'/>}
          {this.state.allBooks.length && <BooksDisplay 
            books={this.filterAllBooks('fantasy')} 
            sectionGenre='Fantasies'/>}
          {this.state.allBooks.length && <BooksDisplay 
            books={this.filterAllBooks('biography')}
            sectionGenre='Biographies' />}
          {this.state.allBooks.length && <BooksDisplay 
            books={this.filterAllBooks('history')} 
            sectionGenre='Histories'/>}
          {this.state.allBooks.length && <BooksDisplay 
            books={this.filterAllBooks('horror')} 
            sectionGenre='Horrors'/>}
          </>
          )

        }} />
        {this.props.currentUser ? <Redirect to='/' /> : null}
    </div>
    )
  }
}

const mapStateToProps = state => ({
  currentUser: state.currentUser,
  favorites: state.favorites,
  searchResults: state.searchResults,
  selectedBook: state.selectedBook
})

const mapDispatchToProps = dispatch => ({
  setCurrentUser: user => dispatch(setCurrentUser(user))
})

export default connect(mapStateToProps, mapDispatchToProps)(App);
