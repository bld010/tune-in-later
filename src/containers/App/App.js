import React, { Component } from 'react';
import BooksDisplay from '../BooksDisplay/BooksDisplay';
import LoginForm from '../LoginForm/LoginForm';
import { landingFetch } from '../../util/apiCalls';
import './App.scss';
import NewUserForm from '../NewUserForm/NewUserForm';
import { Route, NavLink, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { setCurrentUser } from '../../actions';
import SearchForm from '../SearchForm/SearchForm';
import BookDetails from '../BookDetails/BookDetails';
import PropTypes from 'prop-types';

export class App extends Component{
  constructor() {
    super();
    this.state = {
      allBooks : [],
      error: '',
    }
  }
  
  componentDidMount() {
    landingFetch()
      .then(data => this.setState({allBooks: this.handleInitialData(data)}))
      .catch(err => this.setState({error : 'Sorry, there was a problem loading our suggested audiobooks. Please enter a search term to see specific results!'}, () => {console.error('error in landing fetch', err)}))
  }

  handleInitialData = (data) => {
    const allBooks = data.map(datum => datum).flat();
    let regex = new RegExp('(&nbsp;|<([^>]+)>)', 'g')
    const formattedBooks = allBooks.map(book => ({
      artist: book.artistName,
      image: book.artworkUrl100,
      price: book.collectionPrice,
      title: book.collectionName,
      genre: book.primaryGenreName,
      description: book.description.replace(regex, ''),
      date: book.releaseDate,
      filterType: book.filterType,
      id: book.collectionId
    }));
    return formattedBooks
  }

  filterAllBooks = (filterType) => {
    const filteredBooks = this.state.allBooks.filter(book => book.filterType === filterType);
    return filteredBooks;
  }

  render() {
    return (
    <div className="App">
      <header>
        <SearchForm />
        <h1>Listen<span>Later</span></h1>
        <nav>
          <NavLink to='/' className='nav'>Home</NavLink>
          {this.props.currentUser && <NavLink to='/my-collection' className='nav'>My Collection</NavLink>}
          {!this.props.currentUser && <NavLink to='/login' className='nav'>Sign In</NavLink> }
          {this.props.currentUser && <NavLink to='/' className='nav' onClick={() => this.props.setCurrentUser(null)}>Sign Out</NavLink>}
          {/* {this.props.currentUser && !this.props.favorites.length && <h2>You haven't favorited any books yet!</h2>} */}
        </nav>
      </header>
        <Route exact path='/login' render={() => {
          return(
            <>
              <LoginForm />
              <NewUserForm />
            </>)
          }} 
          />
         
      <Route 
        path = '/'
        render={() => {
          return(
            <>
          <section className="main-display">
            {this.props.selectedBook && <BookDetails book={this.props.selectedBook}/>}
            {!this.props.selectedBook && this.props.currentUser ? <h2 className='welcome-message'>Welcome {!this.props.selectedBook && this.props.currentUser.name}</h2> : <h2>Welcome, please sign in!</h2>}
          </section>
          <Route path='/my-collection' render={() => {
            return(
              <>
              <BooksDisplay books={this.props.favorites} sectionGenre='Favorites' />
              </>
            )
          }}
          />
          {(this.props.searchResults.length || null ) && <BooksDisplay 
            books={this.props.searchResults}
            sectionGenre='Search Results' />}
          {(this.state.allBooks.length || null ) && <BooksDisplay 
            books={this.filterAllBooks('romance')} 
            sectionGenre='Romances'/>}
          {(this.state.allBooks.length || null ) && <BooksDisplay 
            books={this.filterAllBooks('fantasy')} 
            sectionGenre='Fantasies'/>}
          {(this.state.allBooks.length || null ) && <BooksDisplay 
            books={this.filterAllBooks('biography')}
            sectionGenre='Biographies' />}
          {(this.state.allBooks.length || null ) && <BooksDisplay 
            books={this.filterAllBooks('history')} 
            sectionGenre='Histories'/>}
          {( this.state.allBooks.length || null ) && <BooksDisplay 
            books={this.filterAllBooks('horror')} 
            sectionGenre='Horrors'/>}
          </>
          )
        }} />
        {this.props.currentUser ? <Redirect to='/' /> : null}
    </div>
    )
  }
}

export const mapStateToProps = state => ({
  currentUser: state.currentUser,
  favorites: state.favorites,
  searchResults: state.searchResults,
  selectedBook: state.selectedBook
})

export const mapDispatchToProps = dispatch => ({
  setCurrentUser: user => dispatch(setCurrentUser(user))
})


export default connect(mapStateToProps, mapDispatchToProps)(App);

App.propTypes = {
  currentUser: PropTypes.object,
  favorites: PropTypes.array.isRequired,
  searchResults: PropTypes.array.isRequired,
  selectedBook: PropTypes.object,
  setCurrentUser: PropTypes.func.isRequired
}
