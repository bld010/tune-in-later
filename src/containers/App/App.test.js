import React from 'react';
import { mapDispatchToProps, App } from './App'
import { shallow } from 'enzyme';
import { setCurrentUser } from '../../actions';

describe('App', () => {

  let wrapper;
  let mockCurrentUser = null;
  let mockFavorites = [];
  let mockSearchResults = [];
  let mockSelectedBook = null;
  let mockBook;
  let mockData;
  let mockFormattedBook;
  let expectedFormattedBooks;


  beforeEach(() => {

    wrapper = shallow(
      <App 
        favorites={mockFavorites}
        currentUser={mockCurrentUser}
        searchResults={mockSearchResults}
        selectedBook={mockSelectedBook}
        setCurrentUser={jest.fn()}
      />
    )

    mockBook = {
      artistName: 'Ron Swanson',
      artworkUrl100: 'bacon.jpg',
      collectionPrice: 200.99,
      collectionName: 'Pyramid of Greatness',
      primaryGenreName: 'History',
      description: 'Eat meat.',
      releaseDate: '1776',
      filterType: 'History',
      collectionId: 994923
    }

    mockData = [
      [mockBook, mockBook],
      [mockBook],
      [mockBook, mockBook, mockBook]
    ]

    mockFormattedBook = {
      artist: 'Ron Swanson',
      image: 'bacon.jpg',
      price: 200.99,
      title: 'Pyramid of Greatness',
      genre: 'History',
      description: 'Eat meat.',
      date: '1776',
      filterType: 'History',
      id: 994923
    }

    expectedFormattedBooks = [
      mockFormattedBook,
      mockFormattedBook,
      mockFormattedBook,
      mockFormattedBook,
      mockFormattedBook,
      mockFormattedBook,    
    ]

  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  describe('handleInitialData', () => {

    it('should format the arrays coming from the fetch', () => {

      
      expect(wrapper.instance().handleInitialData(mockData)).toEqual(expectedFormattedBooks)
    })

  })

  describe('filterAllBooks', () => {

  })

  describe('mapDispatchToProps', () => {
    it('calls dispatch with a setCurrentUser action when setCurrentUser is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = setCurrentUser({name:'Dorbid'})

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setCurrentUser({name:'Dorbid'})

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)

    })
  })


})




