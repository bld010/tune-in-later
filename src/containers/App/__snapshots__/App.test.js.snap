// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div
  className="App"
>
  <header>
    <Connect(SearchForm) />
    <h1>
      ListenLater
    </h1>
    <nav>
      <NavLink
        className="nav"
        to="/"
      >
        Home
      </NavLink>
      <NavLink
        className="nav"
        to="/login"
      >
        Sign In
      </NavLink>
    </nav>
  </header>
  <Route
    exact={true}
    path="/login"
    render={[Function]}
  />
  <Route
    path="/"
    render={[Function]}
  />
</div>
`;
